ddtrace/contrib/anthropic/utils.py

```diff
@@ -0,0 +1,284 @@
+            for content in chunk.message.content:
+                if content.type == "text":
+                    messages.append(
+                        {
+                            "content": content.text,
+                            "role": chunk.message.role,
+                            "finish_reason": chunk.message.stop_reason,
+                            "usage": {
+                                "input": chunk.message.usage.input_tokens,
+                                "output": chunk.message.usage.output_tokens,
+                            },
+                        }
+                    )
+    return messages
+
+
+def _construct_message_from_streamed_chunks(chunk, prev_message=None) -> Tuple[Dict[str, str], bool]:
+    """Constructs a chat message dictionary from streamed chunks.
+    The resulting message dictionary is of form {"content": "...", "role": "...", "finish_reason": "..."}
+    """
+    message = prev_message if prev_message is not None else {"content": ""}
+    message_finished = False
+    if getattr(chunk, "type", "") == "message_start":
+        # this is the starting chunk
+        chunk_content = getattr(chunk.message, "content", "")
+        chunk_role = getattr(chunk.message, "role", "")
+        chunk_usage = getattr(chunk.message, "usage", "")
+        if chunk_content:
+            message["content"] += chunk_content
+        if chunk_role:
+            message["role"] = chunk_role
+        if chunk_usage:
+            message["usage"] = {}
+            message["usage"]["input"] = getattr(chunk_usage, "input_tokens", 0)
+            message["usage"]["output"] = getattr(chunk_usage, "output_tokens", 0)
+    elif getattr(chunk, "delta", None):
+        # delta events contain new content
+        content_block = chunk.delta
+        if getattr(content_block, "type", "") == "text_delta":
+            chunk_content = getattr(content_block, "text", "")
+            if chunk_content:
+                message["content"] += chunk_content
+
+        elif getattr(chunk, "type", "") == "message_delta":
+            # message delta events signal the end of the message
+            chunk_finish_reason = getattr(content_block, "stop_reason", "")
+            if chunk_finish_reason:
+                message["finish_reason"] = content_block.stop_reason
+                message["content"] = message["content"].strip()
+
+                chunk_usage = getattr(chunk, "usage", {})
+                if chunk_usage:
+                    message_usage = message.get("usage", {"output": 0, "input": 0})
+                    message_usage["output"] += getattr(chunk_usage, "output_tokens", 0)
+                    message_usage["input"] += getattr(chunk_usage, "input_tokens", 0)
+                    message["usage"] = message_usage
+                message_finished = True
+
+    return message, message_finished
+
+
+def _tag_streamed_chat_completion_response(integration, span, messages):
```

Comment:
Do Anthropic streamed responses provide `usage` in the output response?

Reply:
yes

Reply:
do we want to tag usage in the integration or the LLMObs integration?